### PR TITLE
register baremetal hosts

### LINKED ISSUE #2063: baremetal:  installer should register hosts for masters and workers
Original issue from @dhellmann here: https://github.com/openshift-metal3/kni-installer/issues/46

---

# Platform:

baremetal

# What happened?

After the cluster was running, there were no BareMetalHost objects associated with the Machines for master Nodes.

# What you expected to happen?

At least the information for the BareMetalHost CRs for the masters should be registered. If those are set up with the machineRef linked to existing Machines, the hosts will be marked as "externally provisioned" and managed more safely (they won't be rebooted to inspect their hardware).

# How to reproduce it (as minimally and precisely as possible)?

```console
$ cd dev-scripts
$ make
```

# Anything else we need to know?

To populate the hardware details for masters we will have to update the Status block of the host objects.

# References

- https://github.com/openshift-metalkube/dev-scripts/pull/330

### PR DESCRIPTION
**Pull request by dhellmann**
Add BareMetalHost assets when registering machines as part of the baremetal platform handling.

This change requires https://github.com/openshift-metal3/dev-scripts/pull/716 to work because the host CRD must be registered during the installation process so the bootstrap scripts do not block trying to register assets generated by the installer.

Closes #2063

### CODE REVIEW

**Comment by dhellmann** [comment]
This depends on https://github.com/openshift/machine-api-operator/pull/302 merging first so that the host CRD is defined when the machine-api-operator is launches the baremetal platform components.

File: pkg/asset/machines/master.go [hunk 1]
```diff
@@ -52,12 +57,20 @@ import (
...
 	MachineConfigFiles []*asset.File
+	SecretFiles        []*asset.File
+	HostFiles          []*asset.File
 	MachineFiles       []*asset.File
 }
 
 const (
 	directory = "openshift"
 
+	// secretFileName is the format string for constucting the Secret filenames.
```
Comment (abhinavdahiya) [inline pkg/asset/machines/master.go]: mispelled `constucting`

File: pkg/asset/machines/master.go [hunk 2]
```diff
@@ -195,10 +212,83 @@ func (m *Master) Generate(dependencies asset.Parents) error {
...
 			return errors.Wrap(err, "failed to create master machine objects")
 		}
+
+		logrus.Debugf("hosts: %v\n", ic.Platform.BareMetal.Hosts)
+
+		// FIXME(dhellmann): Move host creation to a separate
+		// function, as with baremetal.Machines(). Args are probably
+		// the config and the machines (so the hosts and machines can
+		// be matched up).
+		for i, host := range ic.Platform.BareMetal.Hosts {
```
Comment (abhinavdahiya) [inline pkg/asset/machines/master.go]: all this should be in `baremetal` package

**Comment by dhellmann** [comment]
@abhinavdahiya , Thanks for the review. I've submitted this because I can't make the unit test scripts run locally. I will remove the WIP when it's ready for a real review.

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/932/

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/948/

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/959/

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/960/

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/961/

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/965/

**Comment by dhellmann** [comment]
Those failed openstack tests don't have anything to do with this change.

/test e2e-openstack

**Comment by dhellmann** [comment]
@abhinavdahiya I think this is ready for your review now. Thanks!

**Comment by dhellmann** [comment]
I think the KNI CI job failed because of the bad ironic images. That has been resolved, and I have started a new CI job 969

**Comment by abhinavdahiya** [comment]
I think it's best if the comments follow https://github.com/openshift/installer/blob/master/CONTRIBUTING.md#commit-message-format for commit titles.

File: pkg/asset/machines/baremetal/hosts.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+			Data: map[string][]byte{
+				"username": []byte(host.BMC.Username),
+				"password": []byte(host.BMC.Password),
+			},
+		}
+		settings.Secrets = append(settings.Secrets, secret)
+
+		newHost := baremetalhost.BareMetalHost{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "metal3.io/v1alpha1",
```
Comment (abhinavdahiya) [inline pkg/asset/machines/baremetal/hosts.go]: these are best imported from the packages...

File: pkg/asset/machines/baremetal/hosts.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+				BMC: baremetalhost.BMCDetails{
+					Address:         host.BMC.Address,
+					CredentialsName: secret.Name,
+				},
+				BootMACAddress:  host.BootMACAddress,
+				HardwareProfile: host.HardwareProfile,
+			},
+		}
+		if i < len(machines) {
+			// NOTE(dhellmann): Setting ExternallyProvisioned to
```
Comment (abhinavdahiya) [inline pkg/asset/machines/baremetal/hosts.go]: who is this note for? `dhellmann`? or reader of this code?

File: pkg/asset/machines/master.go [hunk 5]
```diff
@@ -52,20 +53,33 @@ import (
 type Master struct {
 	UserDataFile       *asset.File
 	MachineConfigFiles []*asset.File
+	SecretFiles        []*asset.File
```
Comment (abhinavdahiya) [inline pkg/asset/machines/master.go]: I think we should separate this out to a different asset... 

Or alteast make it clear that this is only for bare-metal...

File: pkg/asset/machines/master.go [hunk 6]
```diff
@@ -52,20 +53,33 @@ import (
...
+	HostFiles          []*asset.File
 	MachineFiles       []*asset.File
 }
 
 const (
 	directory = "openshift"
 
-	// masterMachineFileName is the format string for constucting the master Machine filenames.
+	// secretFileName is the format string for constucting the Secret filenames.
```
Comment (abhinavdahiya) [inline pkg/asset/machines/master.go]: These should make it clear that these are only valid for bare-metal.

File: pkg/asset/machines/master.go [hunk 7]
```diff
@@ -195,10 +210,19 @@ func (m *Master) Generate(dependencies asset.Parents) error {
...
 		if err != nil {
 			return errors.Wrap(err, "failed to create master machine objects")
 		}
+
+		hostSettings, err = baremetal.Hosts(ic, machines)
+		if err != nil {
+			return errors.Wrap(err, "failed to assemble host data")
+		}
+
+		logrus.Debugf("hosts: %v\n", hostSettings.Hosts)
```
Comment (abhinavdahiya) [inline pkg/asset/machines/master.go]: please drop the debug.

File: pkg/asset/machines/master.go [hunk 8]
```diff
@@ -109,6 +123,7 @@ func (m *Master) Generate(dependencies asset.Parents) error {
 
 	pool := ic.ControlPlane
 	var err error
+	hostSettings := &baremetal.HostSettings{}
```
Comment (abhinavdahiya) [inline pkg/asset/machines/master.go]: again, make it specific to baremetal..

File: pkg/asset/machines/master.go [hunk 9]
```diff
@@ -243,6 +267,40 @@ func (m *Master) Generate(dependencies asset.Parents) error {
 		}
 	}
 
+	if len(hostSettings.Hosts) > 0 {
```
Comment (abhinavdahiya) [inline pkg/asset/machines/master.go]: why is this not in bare-metal switch??

File: pkg/asset/machines/master.go (hunk 8 above)
Comment (abhinavdahiya) [reply pkg/asset/machines/master.go]: also why is this not in bare-metal switch..

File: pkg/asset/machines/master.go [hunk 10]
```diff
@@ -253,12 +311,28 @@ func (m *Master) Files() []*asset.File {
 		files = append(files, m.UserDataFile)
 	}
 	files = append(files, m.MachineConfigFiles...)
+	// Add the host CRD so we don't block trying to create a type
+	// unknown to the cluster.
+	files = append(files, m.CRDFiles...)
```
Comment (abhinavdahiya) [inline pkg/asset/machines/master.go]: this doesn't seem to be defined... _confused_

File: pkg/asset/machines/master.go [hunk 11]
```diff
@@ -273,12 +347,26 @@ func (m *Master) Load(f asset.FileFetcher) (found bool, err error) {
 		return true, err
 	}
 
-	fileList, err := f.FetchByPattern(filepath.Join(directory, masterMachineFileNamePattern))
+	fileList, err = f.FetchByPattern(filepath.Join(directory, secretFileNamePattern))
 	if err != nil {
 		return true, err
 	}
+	m.SecretFiles = fileList
+	logrus.Debugf("Found %d secret files", len(m.SecretFiles))
```
Comment (abhinavdahiya) [inline pkg/asset/machines/master.go]: please drop the debug

File: pkg/asset/machines/master.go [hunk 12]
```diff
@@ -273,12 +347,26 @@ func (m *Master) Load(f asset.FileFetcher) (found bool, err error) {
...
 
+	fileList, err = f.FetchByPattern(filepath.Join(directory, hostFileNamePattern))
+	if err != nil {
+		return true, err
+	}
+	m.HostFiles = fileList
+	logrus.Debugf("Found %d host files", len(m.HostFiles))
```
Comment (abhinavdahiya) [inline pkg/asset/machines/master.go]: please drop the debug

File: pkg/asset/machines/master.go [hunk 13]
```diff
@@ -253,12 +311,28 @@ func (m *Master) Files() []*asset.File {
...
+
+	for _, file := range files {
+		logrus.Debugf("Master file %s %q", file.Filename, file.Data)
+	}
 	return files
 }
 
 // Load reads the asset files from disk.
 func (m *Master) Load(f asset.FileFetcher) (found bool, err error) {
+	var fileList []*asset.File
```
Comment (abhinavdahiya) [inline pkg/asset/machines/master.go]: define it right above where first required.

File: pkg/asset/machines/master.go [hunk 14]
```diff
@@ -273,12 +347,26 @@ func (m *Master) Load(f asset.FileFetcher) (found bool, err error) {
...
+	}
+	m.HostFiles = fileList
+	logrus.Debugf("Found %d host files", len(m.HostFiles))
+
+	fileList, err = f.FetchByPattern(filepath.Join(directory, masterMachineFileNamePattern))
+	if err != nil {
+		return true, err
+	}
 	m.MachineFiles = fileList
+
```
Comment (abhinavdahiya) [inline pkg/asset/machines/master.go]: extra line..

**Comment by abhinavdahiya** [comment]
Make sure all the bare-metal specific assets being overloaded into master assets are inside the bare-metal switch.

**Comment by metal3ci** [comment]
Build SUCCESS, see build http://10.8.144.11:8080/job/dev-tools/969/

File: pkg/asset/machines/baremetal/hosts.go (hunk 4 above)
Comment (dhellmann) [reply pkg/asset/machines/baremetal/hosts.go]: It's a note by me to the reader of the code. I'm used to including a name like that to aid someone trying to track down the origin of something unusual like this. Should I remove the name?

File: pkg/asset/machines/baremetal/hosts.go (hunk 3 above)
Comment (dhellmann) [reply pkg/asset/machines/baremetal/hosts.go]: ACK

File: pkg/asset/machines/master.go (hunk 5 above)
Comment (dhellmann) [reply pkg/asset/machines/master.go]: I did think about that, but the host assets require information available in the computed Machine assets in order to link the two together and avoid reprovisioning the master hosts.

I couldn't figure out if there is some way to say that when the platform is baremetal we need to add these extra things and make the machines available to the function that computes them. I could say that the Baremetal asset depends on the Machine assets, but it wasn't clear that that wouldn't just end up computing the machines twice. Is that what dependencies.Get() does for us?

What is a good way to make it more clear these are only used for baremetal? Renaming the fields? Adding them to a sub-struct? Comments?

File: pkg/asset/machines/master.go (hunk 8 above)
Comment (dhellmann) [reply pkg/asset/machines/master.go]: This initializes an empty variable. Line 232 gives a populated version. By ensuring the variable always exists, we don't need multiple platform checks before using it.

File: pkg/asset/machines/master.go (hunk 9 above)
Comment (dhellmann) [reply pkg/asset/machines/master.go]: All of the other code to populate the list of files was outside of the switch, so I structured this new code the same way.

File: pkg/asset/machines/master.go (hunk 10 above)
Comment (dhellmann) [reply pkg/asset/machines/master.go]: I'm not sure what isn't defined? m.CRDFiles is defined as part of the Master type.

File: pkg/asset/machines/master.go (hunk 13 above)
Comment (dhellmann) [reply pkg/asset/machines/master.go]: ACK

File: pkg/asset/machines/master.go (hunk 12 above)
Comment (dhellmann) [reply pkg/asset/machines/master.go]: ACK

File: pkg/asset/machines/master.go (hunk 14 above)
Comment (dhellmann) [reply pkg/asset/machines/master.go]: ACK

File: pkg/asset/machines/master.go (hunk 9 above)
Comment (abhinavdahiya) [reply pkg/asset/machines/master.go]: but those are valid for all platforms, these aren't...

File: pkg/asset/machines/master.go (hunk 10 above)
Comment (abhinavdahiya) [reply pkg/asset/machines/master.go]: this change is across 2 commits...

File: pkg/asset/machines/master.go (hunk 5 above)
Comment (abhinavdahiya) [reply pkg/asset/machines/master.go]: can we start by separating them by new line in the struct and commenting...

Comment (dhellmann) [reply pkg/asset/machines/master.go]: Sure, a comment is easy.

File: pkg/asset/machines/master.go (hunk 9 above)
Comment (dhellmann) [reply pkg/asset/machines/master.go]: OK, sure. It looked like the switch statement was for initializing the inputs to build the files, and then the file building happened later. I'll move all of the baremetal file stuff up inside the switch statement.

File: pkg/asset/machines/master.go (hunk 10 above)
Comment (dhellmann) [reply pkg/asset/machines/master.go]: Ah, I must have messed something up when I rebased. I'll fix that.

File: pkg/asset/machines/master.go (hunk 11 above)
Comment (dhellmann) [reply pkg/asset/machines/master.go]: ACK

**Comment by dhellmann** [comment]
@abhinavdahiya I think I've addressed all of your comments in the latest draft.

File: pkg/asset/machines/baremetal/hosts.go [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+// used to construct the cluster.
+func Hosts(config *types.InstallConfig, machines []machineapi.Machine) (*HostSettings, error) {
+	settings := &HostSettings{}
+
+	if config.Platform.BareMetal == nil {
+		return nil, fmt.Errorf("no baremetal platform in configuration")
+	}
+
+	for i, host := range config.Platform.BareMetal.Hosts {
+		logrus.Debugf("configuring host: name = %s, details = %v\n", host.Name, host)
```
Comment (abhinavdahiya) [inline pkg/asset/machines/baremetal/hosts.go]: please drop the debug.

File: pkg/asset/machines/baremetal/hosts.go (hunk 4 above)
Comment (abhinavdahiya) [reply pkg/asset/machines/baremetal/hosts.go]: Yeah, just drop the name, the commit has user defined.

File: pkg/asset/machines/master.go (hunk 7 above)
Comment (abhinavdahiya) [reply pkg/asset/machines/master.go]: This still needs to be dropped.

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/976/

**Comment by metal3ci** [comment]
Build SUCCESS, see build http://10.8.144.11:8080/job/dev-tools/977/

**Comment by dhellmann** [comment]
The latest draft removes the CRD registration work-around.

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/981/

**Comment by dhellmann** [comment]
The dependency has merged and I think this is ready for review again.

**Comment by metal3ci** [comment]
Build SUCCESS, see build http://10.8.144.11:8080/job/dev-tools/993/

**Comment by dhellmann** [comment]
@abhinavdahiya did you have any other review comments on this one? I'd like to get it in by the deadline at the end of the week.

File: pkg/asset/machines/master.go [hunk 16]
```diff
@@ -253,6 +314,13 @@ func (m *Master) Files() []*asset.File {
 		files = append(files, m.UserDataFile)
 	}
 	files = append(files, m.MachineConfigFiles...)
+	// Hosts refer to secrets, so place the secrets before the hosts
+	// to avoid unnecessary reconciliation errors.
```
Comment (wking) [inline pkg/asset/machines/master.go]: Ordering here doesn't matter, does it?  These will be pushed in the order based on their filenames.

File: pkg/asset/machines/master.go [hunk 17]
```diff
@@ -53,19 +53,40 @@ type Master struct {
...
+	HostFiles []*asset.File
 }
 
 const (
 	directory = "openshift"
 
-	// masterMachineFileName is the format string for constucting the master Machine filenames.
+	// secretFileName is the format string for constructing the Secret
+	// filenames for baremetal clusters.
+	secretFileName = "99_openshift-cluster-api_host-bmc-secrets-%s.yaml"
```
Comment (wking) [inline pkg/asset/machines/master.go]: I dunno why we like `99_` so much.  Maybe use `90_` for secrets to put them first and `95_` for hosts?

File: pkg/asset/machines/master.go (hunk 16 above)
Comment (dhellmann) [reply pkg/asset/machines/master.go]: Yeah, I think that comment was from before I fully understood how that part worked and I've just been carrying it along.

File: pkg/asset/machines/master.go (hunk 17 above)
Comment (dhellmann) [reply pkg/asset/machines/master.go]: Sure, I can do that.

Comment (dhellmann) [reply pkg/asset/machines/master.go]: Since this merged, I submitted #2174 to address this comment.

**Pull request merged**